File: python/Figure12.py
# Author: Kenji Kashima
# Date  : 2025/05/25

import numpy as np
import matplotlib.pyplot as plt
import config

np.random.seed(13)

def gaussian_kernel(x,y,c):
    """
    Gaussian kernel function (RBF kernel) in eq 12.7 and 12.8. 

    Parameters:
        x : ndarray
            Input array (e.g., column vector of sample points).
        y : ndarray
            Input array to compare with x (will be transposed inside).
        c : float
            Kernel width (scale parameter).

    Returns:
        ndarray
            Kernel matrix K where K[i,j] = exp( - (x_i - y_j)^2 / c^2 )
    """
    return np.exp(-(x-y.T)**2/c**2)

def min_kernel(x,y):
    """
    Min kernel function in eq. 12.30 and 12.8.

    Parameters:
        x : ndarray
            Input array (1D, will be flattened).
        y : ndarray
            Input array (1D, will be flattened).
    
    Returns:
        ndarray
            Kernel matrix K where K[i,j] = min(x_i, y_j).
    """
    x = np.ravel(x)
    y = np.ravel(y)
    return np.minimum.outer(y, x)

def figure12_2(N_x= 100, c = 0.1,label="a"):
    '''  
        Figure12.2(a) c=0.1 
        Figure12.2(b) c=1.0
    '''
    figsize = config.global_config(type= 1)
    x = np.linspace(0,1,N_x).reshape(-1,1) # x for plot 
    # gaussian kernel function
    K = gaussian_kernel(x,x,c)     

    x = x.flatten()

    # prior distribution
    prior_mean = x # y=mu(x)=x

    plt.figure(figsize=figsize)

    for i in range(5): # 5 samples
         plt.plot(x,np.random.multivariate_normal(prior_mean,K),color="gray",alpha=0.8)  #eq 12.13
    plt.plot(x,prior_mean,label=r"$\mu({\rm x})$")  
    x_SD = np.sqrt(np.diag(K))   # standard deviation of x

    plt.fill_between(x,prior_mean-x_SD,prior_mean+x_SD,color="blue",alpha=0.25)
    plt.xlim([0,1])
    plt.ylim([-3,3])
    plt.xlabel(r"${\rm x}$")
    plt.legend(loc='upper left')
    plt.grid()
    plt.tight_layout()
    plt.savefig("./Figure12_2{}.pdf".format(label))
    plt.show()

def figure12_3(N_x= 100, s_bar = 10, c = 0.1, label="a"):
    '''  
        Figure12.3(a) s_bar=10 
        Figure12.3(b) s_bar=50
    '''
    figsize = config.global_config(type= 1)
    x = np.expand_dims(np.linspace(0,1,N_x),axis=0)
    x_sample = np.random.rand(s_bar) * np.random.rand(s_bar); # x data points
    x_sample = np.expand_dims(x_sample,axis=0)
    
    # gaussian kernel function
    K = gaussian_kernel(x_sample, x_sample, c)  # eq 12.8
    kx = gaussian_kernel(x_sample, x, c)    # eq 12.9

    x = x.flatten()
    x_sample = x_sample.flatten()

    # function in example 12.1.9 and Fig 12.4
    mu_x = lambda x : x                     # mu(x)=x
    fn_x = lambda x :np.sin(4*np.pi*x)    # y(x) = sin(4πx)

    fn_list = fn_x(x)

    sigma = 0.1
    e_s = sigma * np.random.randn(s_bar)    # noise sample
    y_sample = fn_x(x_sample) + e_s         # observation data 

    plt.figure(figsize=figsize)
    plt.plot(x,mu_x(x),color = 'red' ,label=r"$\mu({\rm x})$") 
    plt.plot(x,fn_list,'k--',label=r'$\sin(4\pi {\rm x})$');  

    #plot samples
    plt.scatter(x_sample,y_sample,edgecolor='b',marker='o',facecolor='none', label=r"${\rm y}_s$")

    prior_mean = mu_x(x_sample)
    mean = mu_x(x)+kx @ np.linalg.inv(K + sigma**2 * np.eye(s_bar)) @ (y_sample-prior_mean) # eq 12.16
    Kpost = gaussian_kernel(x, x, c) - kx @ np.linalg.inv(K + sigma**2 * np.eye(s_bar)) @ kx.T  # eq 12.17
    x_SD = np.sqrt(np.diag(Kpost))  # posterior standard deviation

    plt.fill_between(x,mean-x_SD,mean+x_SD,color="blue",alpha=0.25)
    plt.plot(x,mean,"-.", color="blue",   label = r"$\mu ({\rm x}|\mathcal D)$")
    plt.xlim([0,1])
    plt.ylim([-1.5,3.0])
    plt.yticks([-1,0,1,2,3])

    plt.xlabel(r"${\rm x}$")
    plt.legend(loc='upper left')
    plt.grid()
    plt.tight_layout()
    plt.savefig("./Figure12_3{}.pdf".format(label))
    plt.show()
    return (x_sample,y_sample)


def figure12_4(N_x= 100, s_bar = 10, label="a", sample0=None):
    '''  
        Figure12.4(a) s_bar=10 
        Figure12.4(b) s_bar=50
    '''
    figsize = config.global_config(type= 1)
    x=np.expand_dims(np.linspace(0,1,N_x),axis=0)

    # prior distribution
    mu_x = lambda x : x                      # mu(x)=x
    fn_x = lambda x : np.sin(4*np.pi*x)      # y(x) = sin(4πx)
    sigma = 0.1
    if sample0 is not None:
        x_sample,y_sample = sample0
        x_sample = np.expand_dims(x_sample,axis=0)
    else:
        x_sample= np.random.rand(s_bar) * np.random.rand(s_bar); # x data points
        x_sample = np.expand_dims(x_sample,axis=0)

        e_s = sigma * np.random.randn(s_bar)
        y_sample = fn_x(x_sample) + e_s
    
    # gaussian kernel function
    K = min_kernel(x_sample, x_sample) 
    kx = min_kernel(x_sample, x)

    x = x.flatten()
    x_sample = x_sample.flatten()
       
    fn_list = fn_x(x)
    plt.figure(figsize=figsize)
    plt.plot(x,mu_x(x),color = 'red' ,label=r"$\mu({\rm x})$") 
    plt.plot(x,fn_list,'k--',label=r'$\sin(4\pi {\rm x})$');  

    #plot samples
    plt.scatter(x_sample,y_sample,edgecolor='b',marker='o',facecolor='none', label=r"${\rm y}_s$")

    prior_mean = mu_x(x_sample)
    mean = mu_x(x) + kx @ np.linalg.inv(K + sigma**2 * np.eye(s_bar)) @ (y_sample-prior_mean) # eq 12.16
    Kpost = min_kernel(x, x) - kx @ np.linalg.inv(K + sigma**2 * np.eye(s_bar)) @ kx.T  # eq 12.17
    x_SD = np.sqrt(np.diag(Kpost))  # 

    plt.fill_between(x,mean-x_SD,mean+x_SD,color="blue",alpha=0.25)
    plt.plot(x,mean,"-.", color="blue",   label = r"$\mu ({\rm x}|\mathcal D)$")
    plt.xlim([0,1])
    plt.ylim([-1.5,3.0])
    plt.yticks([-1,0,1,2,3])
    plt.xlabel(r"${\rm x}$")
    plt.legend(loc='upper left')
    plt.grid()
    plt.tight_layout()
    plt.savefig("./Figure12_4{}.pdf".format(label))
    plt.show()

if __name__ == '__main__':
    figure12_2(N_x=100, c = 0.1, label="a") #Figure12.2(a)
    figure12_2(N_x=100, c = 1.0, label="b") #Figure12.2(b)
    sample10 = figure12_3(N_x=100, s_bar = 10, label="a") #Figure12.3(a)
    sample50 = figure12_3(N_x=100, s_bar = 50, label="b") #Figure12.3(b)
    figure12_4(N_x=100, s_bar = 10, label="a", sample0 = sample10) #Figure12.4(a)
    figure12_4(N_x=100, s_bar = 50, label="b", sample0 = sample50) #Figure12.4(b)

File: matlab/Figure12.m
% Author: Kenji Kashima
% Date  : 2025/09/01 

seed = 1;

% ===== Figure 12.2 =====
figure12_2(100, 0.1, 'a');  % Figure12.2(a)
figure12_2(100, 1.0, 'b');  % Figure12.2(b)

% ===== Figure 12.3 (Gaussian kernel) =====
figure12_3and4(100, 10, 0.1, 'a', "Gaussian", seed);  % Figure12.3(a)
figure12_3and4(100, 50, 0.1, 'b', "Gaussian", seed);  % Figure12.3(b)

% ===== Figure 12.4 (min kernel) =====
figure12_3and4(100, 10, 0.1, 'a', "min", seed);       % Figure12.4(a)
figure12_3and4(100, 50, 0.1, 'b', "min", seed);       % Figure12.4(b)



% -------------------------------------------------------------------------
function figure12_2(N_x, c, label)

x = linspace(0,1,N_x).';                 % (N_x×1)
K = gaussian_kernel(x, x, c);            % (N_x×N_x)
prior_mean = x;                           % mu(x)=x

Ks = 0.5*(K + K.');                       % 数値的に対称化のみ
[V, D] = eig(Ks, 'vector');               % D: eigenvalues (N_x×1)
D = max(D, 0);                            % 負の微小固有値は 0 に（実数化）
S = V .* sqrt(D.');                       % 相当する「平方根」作用（N_x×N_x）
% y = prior_mean + S * randn(N_x,1);

figure('Name', ['Figure 12.2(' label ')'] ); hold on; grid on;
for i = 1:5
    y = prior_mean + S * randn(N_x,1);
    plot(x, y, 'Color', [0.5 0.5 0.5], 'LineWidth', 1.0, 'HandleVisibility','off');
end
plot(x, prior_mean, 'b-', 'LineWidth', 1.2, 'DisplayName', '$\mu(x)$');

x_SD = sqrt(max(0, diag(K)));
fill([x; flipud(x)], [prior_mean+x_SD; flipud(prior_mean-x_SD)], ...
     [0 0 1], 'FaceAlpha', 0.25, 'EdgeColor', 'none', 'HandleVisibility','off');

xlim([0,1]); ylim([-3,3]);
xlabel('x'); legend('Location','northwest');
title(['Figure 12.2(' label, ')']);
end


% -------------------------------------------------------------------------
function figure12_3and4(N_x, s_bar, c, label, kernel, seed)

rng(seed);

x_data = rand(s_bar,1) .* rand(s_bar,1); % (s_bar×1)
x      = linspace(0,1,N_x).';            % (N_x×1)

% カーネル選択
switch string(kernel)
    case "Gaussian"
        K    = gaussian_kernel(x_data, x_data, c);  % (s_bar×s_bar)
        Ktmp = gaussian_kernel(x,      x,      c);  % (N_x×N_x)
        kx   = gaussian_kernel(x,      x_data, c);  % (N_x×s_bar)
        title_tag = ['3' label];         % Figure12.3(a)/(b)
    case "min"
        K    = min_kernel(x_data, x_data);
        Ktmp = min_kernel(x,      x);
        kx   = min_kernel(x,      x_data);
        title_tag = ['4' label];         % Figure12.4(a)/(b)
    otherwise
        error('Unknown kernel: %s', kernel);
end

% 例 12.1.9 と Fig 12.4 の関数
mu_x = @(v) v;                         % mu(x) = x
fn_x = @(v) sin(4*pi*v);               % y(x) = sin(4πx)

sigma = 0.1;                           % ノイズの標準偏差
e_s = sigma * randn(s_bar,1);          % ノイズ
y_data = fn_x(x_data) + e_s;           % 観測

figure('Name', ['Figure 12.' title_tag] ); hold on; grid on;
plot(x, mu_x(x), 'r-', 'LineWidth', 1.2, 'DisplayName', '$\mu(x)$');
plot(x, fn_x(x), 'k--', 'LineWidth', 1.2, 'DisplayName', '$sin(4\pi x)$');
scatter(x_data, y_data, 36, 'o', 'MarkerEdgeColor', 'b', 'MarkerFaceColor', 'none', ...
        'DisplayName', '$y_s$');

% ===== 事後平均・分散 =====
prior_mean = mu_x(x_data);                 % (s_bar×1)
A = K + sigma^2 * eye(s_bar);              % そのまま（jitter なし）

% alpha = (K + σ²I)^{-1} (y - μ) を inv を使わずに解く
alpha = A \ (y_data - prior_mean);         % (s_bar×1)

% 事後平均: μ(x) + k(x,X) * alpha
post_mean = mu_x(x) + kx * alpha;          % (N_x×1)

% 事後分散の対角: diag(Kxx) - diag(kx * A^{-1} * kx')
% A^{-1} * kx' をまとめて解く
W = A \ (kx.');                            % (s_bar×N_x)

% diag(kx * W) を効率よく計算
second_term_diag = sum(kx .* (W.'), 2);    % (N_x×1)
post_var_diag    = diag(Ktmp) - second_term_diag;
post_SD          = sqrt(max(0, post_var_diag));

% 囲み塗りつぶし ±1SD
fill([x; flipud(x)], [post_mean+post_SD; flipud(post_mean-post_SD)], ...
     [0 0 1], 'FaceAlpha', 0.25, 'EdgeColor', 'none', 'HandleVisibility','off' );

plot(x, post_mean, '-.', 'Color', [0 0 1], 'LineWidth', 1.2, ...
     'DisplayName', '$\mu(x|\mathcal{D})$');

xlim([0,1]); ylim([-1.5, 3.0]);
yticks([-1,0,1,2,3]);
xlabel('x'); legend('Location','northwest');
title(['Figure 12.' title_tag])
end


% -------------------------------------------------------------------------
function K = gaussian_kernel(x, y, c)
% Gaussian (RBF) kernel:
% K_ij = exp(-((x_i - y_j)^2) / c^2)
x = x(:); y = y(:);
K = exp( - ( (x - y.').^2 ) / (c^2) );
end

function K = min_kernel(x, y)
% Min kernel:
% K_ij = min(x_i, y_j)
x = x(:); y = y(:);
K = min(x, y.');
end